### PR TITLE
KubeJS Compat

### PR DESCRIPTION
**Pull request by NovaMachina**

### CODE REVIEW

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/CrucibleRecipeJS.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import novamachina.exnihilosequentia.common.blockentity.crucible.CrucibleTypeEnum;
+
+public class CrucibleRecipeJS extends BaseRecipeJS {
+
+  @Override
+  public void create(ListJS args) {
+    this.inputItems.add(this.parseIngredientItem(args.get(0)));
+    this.crucibleType(args.get(1));
+    this.amount(((Number) args.get(2)).intValue());
+    this.fluidResult((FluidStackJS) args.get(3));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/CrucibleRecipeJS.java]: this is throwing `java.lang.ClassCastException: class java.lang.String cannot be cast to class dev.latvian.mods.kubejs.fluid.FluidStackJS`
The Problem here is I add a the fluid `"minecraft:water"` but it doesn't work as this should be converted in a FluidStack

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidItemRecipeJS.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import dev.latvian.mods.kubejs.fluid.FluidStackJS;
+import dev.latvian.mods.kubejs.util.ListJS;
+
+public class FluidItemRecipeJS extends BaseRecipeJS {
+
+  @Override
+  public void create(ListJS args) {
+    this.inputItems.add(this.parseIngredientItem(args.get(0)));
+    this.outputItems.add(this.parseResultItem(args.get(1)));
+    this.fluid((FluidStackJS) args.get(2));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidItemRecipeJS.java]: this is throwing java.lang.ClassCastException: class java.lang.String cannot be cast to class dev.latvian.mods.kubejs.fluid.FluidStackJS
The Problem here is I add the fluid "exnihilosequentia:witch_water" but it doesn't work as this should be converted in a FluidStack

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidOnTopRecipeJS.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.google.gson.JsonObject;
+import dev.latvian.mods.kubejs.fluid.FluidStackJS;
+import dev.latvian.mods.kubejs.util.ListJS;
+
+public class FluidOnTopRecipeJS extends BaseRecipeJS {
+
+  @Override
+  public void create(ListJS args) {
+    this.fluidInTank((FluidStackJS) args.get(0));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidOnTopRecipeJS.java]: this is throwing java.lang.ClassCastException: class java.lang.String cannot be cast to class dev.latvian.mods.kubejs.fluid.FluidStackJS
The Problem here is I add the fluid "exnihilosequentia:witch_water" but it doesn't work as this should be converted in a FluidStack

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidOnTopRecipeJS.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    this.fluidOnTop((FluidStackJS) args.get(1));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidOnTopRecipeJS.java]: this is throwing java.lang.ClassCastException: class java.lang.String cannot be cast to class dev.latvian.mods.kubejs.fluid.FluidStackJS
The Problem here is I add the fluid "exnihilosequentia:sea_water" but it doesn't work as this should be converted in a FluidStack

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidTransformRecipeJS.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import com.google.gson.JsonObject;
+import dev.latvian.mods.kubejs.fluid.FluidStackJS;
+import dev.latvian.mods.kubejs.item.ingredient.IngredientJS;
+import dev.latvian.mods.kubejs.util.ListJS;
+
+public class FluidTransformRecipeJS extends BaseRecipeJS {
+
+  @Override
+  public void create(ListJS args) {
+    this.fluidInTank((FluidStackJS) args.get(0));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidTransformRecipeJS.java]: this is throwing java.lang.ClassCastException: class java.lang.String cannot be cast to class dev.latvian.mods.kubejs.fluid.FluidStackJS
The Problem here is I add the fluid "minecraft:lava" but it doesn't work as this should be converted in a FluidStack

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidTransformRecipeJS.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    this.result((FluidStackJS) args.get(1));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/FluidTransformRecipeJS.java]: this is throwing java.lang.ClassCastException: class java.lang.String cannot be cast to class dev.latvian.mods.kubejs.fluid.FluidStackJS
The Problem here is I add the fluid "minecraft:water" but it doesn't work as this should be converted in a FluidStack

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/SieveRecipeJS.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  public void create(ListJS args) {
+    this.inputItems.add(this.parseIngredientItem(args.get(0)));
+    this.outputItems.add(this.parseResultItem(args.get(1)));
+  }
+
+  public SieveRecipeJS addRoll(float chance, String mesh) {
+    if (!isValidMesh(mesh)) {
+      throw new RecipeExceptionJS("Mesh type must be " + MeshType.printList());
+    }
+    if (this.json.has("rolls")) {
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/SieveRecipeJS.java:25]: maybe add a throw new RecipeExceptionJS with focus on chance not greater than 1. it shows up as in my example 200% but just drops 1 (same as crafttweaker)

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/HeatRecipeJS.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import dev.latvian.mods.kubejs.util.ListJS;
+import dev.latvian.mods.kubejs.util.MapJS;
+
+public class HeatRecipeJS extends BaseRecipeJS {
+
+  @Override
+  public void create(ListJS args) {
+    this.source(ItemStackJS.of(args.get(0)));
+    this.amount(((Number) args.get(1)).intValue());
+    this.properties((MapJS)args.get(2));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/HeatRecipeJS.java]: i first had trouble making a heat recipe, but after adding `{}` as empty additional properties, it works

**Review by remplerus — CHANGES_REQUESTED** [review]
most errors from getting fluids into kubejs, added markers to those errors.
some QoL changes marked.
crooking grass block in my example does drop additional dirt blocks. probably a problem in `common\loot\modifier\UseCrookModifier`

Here's my script added to `kubejs\server_scripts\script.js`:
`
onEvent('recipes', event => {
	event.recipes.exnihilosequentia.compost('minecraft:cobblestone', 750)
	event.recipes.exnihilosequentia.crook('minecraft:grass_block').addDrop('minecraft:coal', 4, 0.5).addDrop('minecraft:iron_ingot', 1, 0.25)
	event.recipes.exnihilosequentia.crucible('minecraft:dirt', 'fired', 500, 'minecraft:water')
	event.recipes.exnihilosequentia.fluid_item('minecraft:stone', 'minecraft:stone_bricks', 'exnihilosequentia:witch_water')
	event.recipes.exnihilosequentia.fluid_on_top('exnihilosequentia:witch_water', 'exnihilosequentia:sea_water', 'minecraft:coarse_dirt')
	event.recipes.exnihilosequentia.fluid_transform('minecraft:lava', 'minecraft:water', '#forge:ores')
	event.recipes.exnihilosequentia.hammer('minecraft:pumpkin').addResult('minecraft:melon_slice', 20, 0.75).addResult('minecraft:pumpkin_seeds', 1, 1)
	event.recipes.exnihilosequentia.heat('minecraft:hay_block', 200, {})
	event.recipes.exnihilosequentia.sieve('minecraft:coarse_dirt', 'minecraft:cobblestone').addRoll(2, 'iron')
})
`

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/SimpleChanceRecipeJS.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.google.gson.JsonArray;
+import dev.latvian.mods.kubejs.item.ItemStackJS;
+import dev.latvian.mods.kubejs.util.ListJS;
+
+public class SimpleChanceRecipeJS extends BaseRecipeJS {
+
+  @Override
+  public void create(ListJS args) {
+    this.inputItems.add(this.parseIngredientItem(args.get(0)));
```
Comment (remplerus) [inline src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/SimpleChanceRecipeJS.java:11]: i have to add something here too: we need to add those blocks to their equivalent BlockTag ("mineable/crook" and "mineable/hammer")

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/HeatRecipeJS.java (hunk 8 above)
Comment (NovaMachina) [reply src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/HeatRecipeJS.java]: I'll make this an optional parameter and then make sure to add this to the documentation.

File: src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/CrucibleRecipeJS.java (hunk 1 above)
Comment (NovaMachina) [reply src/main/java/novamachina/exnihilosequentia/common/compat/kubejs/CrucibleRecipeJS.java]: Fixed

**Review by remplerus — APPROVED** [review]

**Pull request merged**